File: 05.3-api-optimize/src/main/java/spark/basic/MapSideJoin.java
package spark.basic;

import org.apache.spark.api.java.JavaSparkContext;
import org.apache.spark.broadcast.Broadcast;
import org.apache.spark.sql.Dataset;
import org.apache.spark.sql.Row;
import org.apache.spark.sql.SparkSession;


/**
 * Concept:
 * 
 * In general, since your data are distributed among many nodes, 
 * they have to be shuffled before a join that causes significant network I/O and slow performance. 
 * 
 * In cases if we need to join large tables(fact tables) with smaller ones(dimension tables),
 * instead of sending large table data over network, we can just broadcast the smaller table to
 * all nodes to perform a map-side/broadcast join.
 * 
 * Using broadcast variables which can be shared across nodes 
 * ------------------------------------------IMPORTANT----------------------------------------
 * Samples Used:(In this project directory)
 * 1. store_locations.json
 * 2. us_states.json
 *---------------------------------------------------------------------------------------------
 */
public class MapSideJoin {

	public static void main(String[] args) {
		SparkSession spark = SparkSession.builder()
				.appName("Spark - MapSideJoin")
				.master("local[*]")
				.getOrCreate();

		JavaSparkContext sparkContext = new JavaSparkContext(spark.sparkContext());
		System.out.println(sparkContext.getConf().toDebugString());

		//Dataframe containing US states details 
		Dataset<Row> stateDf = spark.read()
				.json("/data/joins/us_states.json");

		// Create a DataFrame based on the store locations.
		Dataset<Row> storesDF = spark.read()
				.json("/data/joins/store_locations.json");

		//## TODO 1 - Set stateDf as a broadcast variable so that,
		//it will be shared across the nodes in the cluster
		//hint broadcast method available in sparkcontext
		// Create a DataFrame of US state data with the broadcast variables.
		
		//Broadcast<???> stateDFBroadcasted = ??????;

		// Join the DataFrames to get an aggregate count of stores in each US Region
		System.out.println("###Number of stores in each US region :");
		/*Dataset<Row> joinedDF = storesDF.join(stateDFBroadcasted.value(), "state")
				.groupBy("census_region")
				.count();		
		joinedDF.show();*/
		spark.close();
		sparkContext.close();
	}
}

File: 05-api/python/mapsidejoin.py
from pyspark.sql import SparkSession
from pyspark.sql.functions import broadcast

spark = SparkSession.builder.appName("spark - MapSideJoin").getOrCreate()
sc = spark.sparkContext

# Concept:

# In general, since your data are distributed among many nodes, 
# they have to be shuffled before a join that causes significant network I/O and slow performance. 

# In cases if we need to join large tables(fact tables) with smaller ones(dimension tables),
# instead of sending large table data over network, we can just broadcast the smaller table to
# all nodes to perform a map-side/broadcast join.

# Using broadcast variables which can be shared across nodes 
# ------------------------------------------IMPORTANT----------------------------------------
# Samples Used:(In this project directory)
# 1. store_locations.json
# 2. us_states.json
#  *---------------------------------------------------------------------------------------------

states_df = spark.read.json('/data/joins/us_states.json')
stores_df = spark.read.json('/data/joins/store_locations.json')

# ## TODO 1 - Set stateDf as a broadcast variable so that,
#  it will be shared across the nodes in the cluster
# Hint broadcast method available in sparkcontext
# Create a DataFrame of US state data with the broadcast variables.

# states_df_broadcasted = ??????????????????

# More hint: 
# states_df_broadcasted = broadcast(states_df)

states_df_broadcasted = broadcast(states_df)

# Join the DataFrames to get an aggregate count of stores in each US Region
print("###Number of stores in each US region :")

joined_df = stores_df.join(states_df_broadcasted, "state").groupBy("census_region").count();		

joined_df.show()

print("Press Enter to continue...")

scanner = sc._gateway.jvm.java.util.Scanner
sys_in = getattr(sc._gateway.jvm.java.lang.System, 'in')
result = scanner(sys_in).nextLine()

spark.stop()

